#include "templateeditor.h"
#include "ui_templateeditor.h"
#include "styleselectordialog.h"
#include "appsettings.h"
#include <QFileDialog>
#include <QTableWidgetItem>
#include <QDialog>
#include <QTextBrowser>
#include <QVBoxLayout>
#include <QDialogButtonBox>
#include <QScrollArea>
#include <QWidget>
#include <QGridLayout>
#include <QLabel>
#include <QLineEdit>
#include <QRegularExpression>
#include <QTextDocumentFragment>
#include <QApplication>
#include <QClipboard>
#include <QPropertyAnimation>


class ClickableLabel : public QLabel {
public:
    explicit ClickableLabel(const QString& text, QWidget* parent = nullptr) : QLabel(text, parent) {
        setCursor(Qt::PointingHandCursor);
        setStyleSheet("QLabel { background-color: #2d2d2d; border: 1px solid #ccc; border-radius: 4px; padding: 2px; }");
        setToolTip("–ù–∞–∂–º–∏—Ç–µ, —á—Ç–æ–±—ã —Å–∫–æ–ø–∏—Ä–æ–≤–∞—Ç—å");
    }

protected:
    void mousePressEvent(QMouseEvent *event) override {
        QApplication::clipboard()->setText(this->text());

        QPropertyAnimation *animation = new QPropertyAnimation(this, "styleSheet", this);
        animation->setDuration(500);
        animation->setStartValue("QLabel { background-color: #4f4f4f; border: 1px solid #5a9e5a; border-radius: 4px; padding: 2px; }"); // –ó–µ–ª–µ–Ω—ã–π
        animation->setEndValue("QLabel { background-color: #2d2d2d; border: 1px solid #ccc; border-radius: 4px; padding: 2px; }");   // –ò—Å—Ö–æ–¥–Ω—ã–π

        QLabel::mousePressEvent(event);
    }
};

TemplateEditor::TemplateEditor(QWidget *parent) :
    QDialog(parent),
    ui(new Ui::TemplateEditor)
{
    ui->setupUi(this);

    connect(ui->addSubstitutionButton, &QPushButton::clicked, this, [this](){
        int row = ui->substitutionsTableWidget->rowCount();
        ui->substitutionsTableWidget->insertRow(row);
        ui->substitutionsTableWidget->setItem(row, 0, new QTableWidgetItem("–¢–µ–∫—Å—Ç –¥–ª—è –ø–æ–∏—Å–∫–∞"));
        ui->substitutionsTableWidget->setItem(row, 1, new QTableWidgetItem("–¢–µ–∫—Å—Ç –¥–ª—è –∑–∞–º–µ–Ω—ã"));
    });

    connect(ui->removeSubstitutionButton, &QPushButton::clicked, this, [this](){
        int currentRow = ui->substitutionsTableWidget->currentRow();
        if (currentRow >= 0) {
            ui->substitutionsTableWidget->removeRow(currentRow);
        }
    });

    connect(ui->browsePosterButton, &QPushButton::clicked, this, [this](){
        QString path = QFileDialog::getOpenFileName(this, "–í—ã–±–µ—Ä–∏—Ç–µ —Ñ–∞–π–ª –ø–æ—Å—Ç–µ—Ä–∞", "", "–ò–∑–æ–±—Ä–∞–∂–µ–Ω–∏—è (*.png *.jpg *.jpeg *.webp)");
        if (!path.isEmpty()) {
            ui->posterPathEdit->setText(path);
        }
    });

    connect(ui->buttonBox, &QDialogButtonBox::accepted, this, &TemplateEditor::accept);
    connect(ui->buttonBox, &QDialogButtonBox::rejected, this, &TemplateEditor::reject);
}

TemplateEditor::~TemplateEditor()
{
    delete ui;
}

void TemplateEditor::setTemplate(const ReleaseTemplate &t)
{
    // –í–∫–ª–∞–¥–∫–∞ "–û—Å–Ω–æ–≤–Ω—ã–µ"
    ui->templateNameEdit->setText(t.templateName);
    ui->seriesTitleEdit->setText(t.seriesTitle);
    ui->releaseTagsEdit->setText(t.releaseTags.join(", "));
    ui->rssUrlEdit->setText(t.rssUrl.toString());
    ui->animationStudioEdit->setText(t.animationStudio);
    ui->subAuthorEdit->setText(t.subAuthor);
    ui->originalLanguageEdit->setText(t.originalLanguage);
    ui->targetAudioFormatComboBox->setCurrentText(t.targetAudioFormat);
    ui->createSrtMasterCheckBox->setChecked(t.createSrtMaster);
    ui->isCustomTranslationCheckBox->setChecked(t.isCustomTranslation);
    ui->useConcatRenderCheckBox->setChecked(t.useConcatRender);
    ui->renderPresetComboBox->clear();
    for(const auto& preset : AppSettings::instance().renderPresets()) {
        ui->renderPresetComboBox->addItem(preset.name);
    }
    ui->renderPresetComboBox->setCurrentText(t.renderPresetName);

    // –í–∫–ª–∞–¥–∫–∞ "–°–æ–∑–¥–∞–Ω–∏–µ –¢–ë"
    ui->generateTbCheckBox->setChecked(t.generateTb);
    ui->endingChapterNameEdit->setText(t.endingChapterName);
    ui->endingStartTimeEdit->setTime(QTime::fromString(t.endingStartTime, "H:mm:ss.zzz"));
    ui->useManualTimeCheckBox->setChecked(t.useManualTime);
    ui->defaultTbStyleComboBox->clear();
    for(const auto& style : AppSettings::instance().tbStyles()) {
        ui->defaultTbStyleComboBox->addItem(style.name);
    }
    ui->defaultTbStyleComboBox->setCurrentText(t.defaultTbStyleName);
    if (t.voiceoverType == ReleaseTemplate::VoiceoverType::Voiceover) {
        ui->voiceoverTypeComboBox->setCurrentText("–ó–∞–∫–∞–¥—Ä");
    } else {
        ui->voiceoverTypeComboBox->setCurrentText("–î—É–±–ª—è–∂");
    }
    ui->castEdit->setPlainText(t.cast.join(", "));
    ui->directorEdit->setText(t.director);
    ui->assistantDirectorEdit->setText(t.assistantDirector);
    ui->soundEngineerEdit->setText(t.soundEngineer);
    ui->songsSoundEngineerEdit->setText(t.songsSoundEngineer);
    ui->episodeSoundEngineerEdit->setText(t.episodeSoundEngineer);
    ui->recordingSoundEngineerEdit->setText(t.recordingSoundEngineer);
    ui->videoLocalizationAuthorEdit->setText(t.videoLocalizationAuthor);
    ui->timingAuthorEdit->setText(t.timingAuthor);
    ui->signsAuthorEdit->setText(t.signsAuthor);
    ui->translationEditorEdit->setText(t.translationEditor);
    ui->releaseBuilderEdit->setText(t.releaseBuilder);

    // –í–∫–ª–∞–¥–∫–∞ "–°—É–±—Ç–∏—Ç—Ä—ã"
    ui->sourceHasSubtitlesCheckBox->setChecked(t.sourceHasSubtitles);
    ui->forceSignStyleRequestCheckBox->setChecked(t.forceSignStyleRequest);
    ui->signStylesEdit->setPlainText(t.signStyles.join('\n'));
    ui->pauseForSubEditCheckBox->setChecked(t.pauseForSubEdit);
    ui->substitutionsTableWidget->setRowCount(0);
    for (auto it = t.substitutions.constBegin(); it != t.substitutions.constEnd(); ++it) {
        int row = ui->substitutionsTableWidget->rowCount();
        ui->substitutionsTableWidget->insertRow(row);
        ui->substitutionsTableWidget->setItem(row, 0, new QTableWidgetItem(it.key()));
        ui->substitutionsTableWidget->setItem(row, 1, new QTableWidgetItem(it.value()));
    }

    // –í–∫–ª–∞–¥–∫–∞ "–®–∞–±–ª–æ–Ω—ã –ø–æ—Å—Ç–æ–≤"
    ui->postTgMp4Edit->setPlainText(t.postTemplates.value("tg_mp4"));
    ui->postTgMkvEdit->setPlainText(t.postTemplates.value("tg_mkv"));
    ui->postVkEdit->setPlainText(t.postTemplates.value("vk"));
    ui->postVkCommentEdit->setPlainText(t.postTemplates.value("vk_comment"));

    // –í–∫–ª–∞–¥–∫–∞ "–ü—É–±–ª–∏–∫–∞—Ü–∏—è"
    ui->seriesTitleForPostEdit->setText(t.seriesTitleForPost);
    ui->totalEpisodesSpinBox->setValue(t.totalEpisodes);
    ui->posterPathEdit->setText(t.posterPath);
    ui->uploadUrlsEdit->setPlainText(t.uploadUrls.join('\n'));
}

ReleaseTemplate TemplateEditor::getTemplate() const
{
    ReleaseTemplate t;

    // –í–∫–ª–∞–¥–∫–∞ "–û—Å–Ω–æ–≤–Ω—ã–µ"
    t.templateName = ui->templateNameEdit->text().trimmed();
    t.seriesTitle = ui->seriesTitleEdit->text().trimmed();
    QStringList tags = ui->releaseTagsEdit->text().split(',', Qt::SkipEmptyParts);
    for(QString& tag : tags) {
        t.releaseTags.append(tag.trimmed());
    }
    t.rssUrl = QUrl(ui->rssUrlEdit->text().trimmed());
    t.animationStudio = ui->animationStudioEdit->text().trimmed();
    t.subAuthor = ui->subAuthorEdit->text().trimmed();
    t.originalLanguage = ui->originalLanguageEdit->text().trimmed();
    t.targetAudioFormat = ui->targetAudioFormatComboBox->currentText();
    t.createSrtMaster = ui->createSrtMasterCheckBox->isChecked();
    t.isCustomTranslation = ui->isCustomTranslationCheckBox->isChecked();
    t.useConcatRender = ui->useConcatRenderCheckBox->isChecked();
    t.renderPresetName = ui->renderPresetComboBox->currentText();

    // –í–∫–ª–∞–¥–∫–∞ "–°–æ–∑–¥–∞–Ω–∏–µ –¢–ë"
    t.generateTb = ui->generateTbCheckBox->isChecked();
    t.endingChapterName = ui->endingChapterNameEdit->text().trimmed();
    t.endingStartTime = ui->endingStartTimeEdit->time().toString("H:mm:ss.zzz");
    t.useManualTime = ui->useManualTimeCheckBox->isChecked();
    t.defaultTbStyleName = ui->defaultTbStyleComboBox->currentText();
    if (ui->voiceoverTypeComboBox->currentText() == "–ó–∞–∫–∞–¥—Ä") {
        t.voiceoverType = ReleaseTemplate::VoiceoverType::Voiceover;
    } else {
        t.voiceoverType = ReleaseTemplate::VoiceoverType::Dubbing;
    }
    QRegularExpression rx("((\\, )|\\,|\\n)");
    t.cast = ui->castEdit->toPlainText().split(rx, Qt::SkipEmptyParts);
    t.director = ui->directorEdit->text().trimmed();
    t.assistantDirector = ui->assistantDirectorEdit->text().trimmed();
    t.soundEngineer = ui->soundEngineerEdit->text().trimmed();
    t.songsSoundEngineer = ui->songsSoundEngineerEdit->text().trimmed();
    t.episodeSoundEngineer = ui->episodeSoundEngineerEdit->text().trimmed();
    t.recordingSoundEngineer = ui->recordingSoundEngineerEdit->text().trimmed();
    t.videoLocalizationAuthor = ui->videoLocalizationAuthorEdit->text().trimmed();
    t.timingAuthor = ui->timingAuthorEdit->text().trimmed();
    t.signsAuthor = ui->signsAuthorEdit->text().trimmed();
    t.translationEditor = ui->translationEditorEdit->text().trimmed();
    t.releaseBuilder = ui->releaseBuilderEdit->text().trimmed();

    // –í–∫–ª–∞–¥–∫–∞ "–°—É–±—Ç–∏—Ç—Ä—ã"
    t.sourceHasSubtitles = ui->sourceHasSubtitlesCheckBox->isChecked();
    t.forceSignStyleRequest = ui->forceSignStyleRequestCheckBox->isChecked();
    t.signStyles = ui->signStylesEdit->toPlainText().split('\n', Qt::SkipEmptyParts);
    t.pauseForSubEdit = ui->pauseForSubEditCheckBox->isChecked();
    t.substitutions.clear();
    for (int row = 0; row < ui->substitutionsTableWidget->rowCount(); ++row) {
        QString findText = ui->substitutionsTableWidget->item(row, 0)->text();
        QString replaceText = ui->substitutionsTableWidget->item(row, 1)->text();
        if (!findText.isEmpty()) {
            t.substitutions.insert(findText, replaceText);
        }
    }

    // –í–∫–ª–∞–¥–∫–∞ "–®–∞–±–ª–æ–Ω—ã –ø–æ—Å—Ç–æ–≤"
    t.postTemplates.clear();
    t.postTemplates["tg_mp4"] = ui->postTgMp4Edit->toPlainText();
    t.postTemplates["tg_mkv"] = ui->postTgMkvEdit->toPlainText();
    t.postTemplates["vk"] = ui->postVkEdit->toPlainText();
    t.postTemplates["vk_comment"] = ui->postVkCommentEdit->toPlainText();

    // –í–∫–ª–∞–¥–∫–∞ "–ü—É–±–ª–∏–∫–∞—Ü–∏—è"
    t.seriesTitleForPost = ui->seriesTitleForPostEdit->text();
    t.totalEpisodes = ui->totalEpisodesSpinBox->value();
    t.posterPath = ui->posterPathEdit->text();
    t.uploadUrls = ui->uploadUrlsEdit->toPlainText().split('\n', Qt::SkipEmptyParts);

    return t;
}

void TemplateEditor::on_selectStylesButton_clicked()
{
    QString filePath = QFileDialog::getOpenFileName(this, "–í—ã–±–µ—Ä–∏—Ç–µ .ass —Ñ–∞–π–ª –¥–ª—è –∞–Ω–∞–ª–∏–∑–∞", "", "ASS Subtitles (*.ass)");
    if (filePath.isEmpty()) return;

    StyleSelectorDialog dialog(this);
    dialog.analyzeFile(filePath);

    if (dialog.exec() == QDialog::Accepted) {
        QStringList selectedStyles = dialog.getSelectedStyles();
        ui->signStylesEdit->setPlainText(selectedStyles.join('\n'));
    }
}

void TemplateEditor::on_helpButton_clicked()
{
    static QDialog* helpDialog = nullptr;
    if (helpDialog && helpDialog->isVisible()) {
        helpDialog->activateWindow();
        return;
    }

    helpDialog = new QDialog(this);
    helpDialog->setAttribute(Qt::WA_DeleteOnClose);
    helpDialog->setWindowTitle("–°–ø—Ä–∞–≤–∫–∞ –ø–æ —à–∞–±–ª–æ–Ω–∞–º –∏ —Ñ–æ—Ä–º–∞—Ç–∏—Ä–æ–≤–∞–Ω–∏—é");
    helpDialog->setMinimumSize(500, 800);

    QScrollArea *scrollArea = new QScrollArea(helpDialog);
    scrollArea->setWidgetResizable(true);
    scrollArea->setStyleSheet("QScrollArea { border: none; }");
    QWidget *scrollWidget = new QWidget();
    scrollArea->setWidget(scrollWidget);
    QVBoxLayout *mainLayout = new QVBoxLayout(scrollWidget);

    auto addRow = [&](QGridLayout* layout, int row, const QString& code, const QString& description) {
        layout->addWidget(new ClickableLabel(code), row, 0);
        layout->addWidget(new QLabel(description), row, 1);
    };

    // --- –ë–ª–æ–∫ 1: –ü–ª–µ–π—Å—Ö–æ–ª–¥–µ—Ä—ã ---
    mainLayout->addWidget(new QLabel("<h3>–ü–ª–µ–π—Å—Ö–æ–ª–¥–µ—Ä—ã (–Ω–∞–∂–º–∏—Ç–µ, —á—Ç–æ–±—ã —Å–∫–æ–ø–∏—Ä–æ–≤–∞—Ç—å):</h3>"));
    QWidget* placeholdersWidget = new QWidget();
    QGridLayout* placeholdersLayout = new QGridLayout(placeholdersWidget);
    addRow(placeholdersLayout, 0, "%SERIES_TITLE%", "–ù–∞–∑–≤–∞–Ω–∏–µ —Å–µ—Ä–∏–∞–ª–∞ –¥–ª—è –ø–æ—Å—Ç–æ–≤");
    addRow(placeholdersLayout, 1, "%EPISODE_NUMBER%", "–ù–æ–º–µ—Ä —Ç–µ–∫—É—â–µ–π —Å–µ—Ä–∏–∏");
    addRow(placeholdersLayout, 2, "%TOTAL_EPISODES%", "–û–±—â–µ–µ –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ —Å–µ—Ä–∏–π");
    addRow(placeholdersLayout, 3, "%CAST_LIST%", "–°–ø–∏—Å–æ–∫ –∞–∫—Ç–µ—Ä–æ–≤ —á–µ—Ä–µ–∑ –∑–∞–ø—è—Ç—É—é");
    addRow(placeholdersLayout, 4, "%DIRECTOR%", "–†–µ–∂–∏—Å—Å–µ—Ä –¥—É–±–ª—è–∂–∞");
    addRow(placeholdersLayout, 5, "%SOUND_ENGINEER%", "–ó–≤—É–∫–æ—Ä–µ–∂–∏—Å—Å–µ—Ä");
    addRow(placeholdersLayout, 6, "%SONG_ENGINEER%", "–ó–≤—É–∫–æ—Ä–µ–∂–∏—Å—Å–µ—Ä –ø–µ—Å–µ–Ω");
    addRow(placeholdersLayout, 7, "%EPISODE_ENGINEER%", "–ó–≤—É–∫–æ—Ä–µ–∂–∏—Å—Å–µ—Ä —ç–ø–∏–∑–æ–¥–∞");
    addRow(placeholdersLayout, 8, "%RECORDING_ENGINEER%", "–ó–≤—É–∫–æ—Ä–µ–∂–∏—Å—Å–µ—Ä –∑–∞–ø–∏—Å–∏");
    addRow(placeholdersLayout, 9, "%SUB_AUTHOR%", "–ê–≤—Ç–æ—Ä –ø–µ—Ä–µ–≤–æ–¥–∞");
    addRow(placeholdersLayout, 10, "%TIMING_AUTHOR%", "–†–∞–∑–º–µ—Ç–∫–∞ (—Ç–∞–π–º–∏–Ω–≥)");
    addRow(placeholdersLayout, 11, "%SIGNS_AUTHOR%", "–õ–æ–∫–∞–ª–∏–∑–∞—Ü–∏—è –Ω–∞–¥–ø–∏—Å–µ–π");
    addRow(placeholdersLayout, 12, "%TRANSLATION_EDITOR%", "–†–µ–¥–∞–∫—Ç–æ—Ä –ø–µ—Ä–µ–≤–æ–¥–∞");
    addRow(placeholdersLayout, 13, "%RELEASE_BUILDER%", "–°–±–æ—Ä–∫–∞ —Ä–µ–ª–∏–∑–∞");
    addRow(placeholdersLayout, 14, "%LINK_ANILIB%", "–°—Å—ã–ª–∫–∞ –Ω–∞ Anilib (–∏–∑ –ø–∞–Ω–µ–ª–∏ '–ü—É–±–ª–∏–∫–∞—Ü–∏—è')");
    addRow(placeholdersLayout, 15, "%LINK_ANIME365%", "–°—Å—ã–ª–∫–∞ –Ω–∞ Anime365 (–∏–∑ –ø–∞–Ω–µ–ª–∏ '–ü—É–±–ª–∏–∫–∞—Ü–∏—è')");
    placeholdersLayout->setColumnStretch(1, 1);
    mainLayout->addWidget(placeholdersWidget);

    mainLayout->addWidget(new QFrame);

    // --- –ë–ª–æ–∫ 2: –§–æ—Ä–º–∞—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Telegram ---
    mainLayout->addWidget(new QLabel("<h3>–§–æ—Ä–º–∞—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ Telegram (–ø—Å–µ–≤–¥–æ-Markdown):</h3>"));
    QTextBrowser* tgHelpBrowser = new QTextBrowser();
    tgHelpBrowser->setOpenExternalLinks(true);
    tgHelpBrowser->setHtml(R"(
        <p>–î–ª—è —Ñ–æ—Ä–º–∞—Ç–∏—Ä–æ–≤–∞–Ω–∏—è —Ç–µ–∫—Å—Ç–∞ –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ —Å–ø–µ—Ü–∏–∞–ª—å–Ω—ã–µ —Å–∏–º–≤–æ–ª—ã. –ü—Ä–∏ –∫–æ–ø–∏—Ä–æ–≤–∞–Ω–∏–∏ —Ç–µ–∫—Å—Ç–∞ –¥–ª—è Telegram, –æ–Ω –±—É–¥–µ—Ç –∞–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏ –ø—Ä–µ–æ–±—Ä–∞–∑–æ–≤–∞–Ω –≤ –Ω—É–∂–Ω—ã–π —Ñ–æ—Ä–º–∞—Ç.</p>
        <h4>–ë–∞–∑–æ–≤—ã–µ —Å—Ç–∏–ª–∏</h4>
        <table border="1" cellspacing="0" cellpadding="5">
            <tr><th>–°—Ç–∏–ª—å</th><th>–°–∏–Ω—Ç–∞–∫—Å–∏—Å</th></tr>
            <tr><td><b>–ñ–∏—Ä–Ω—ã–π</b></td><td><code>**–ñ–∏—Ä–Ω—ã–π —Ç–µ–∫—Å—Ç**</code></td></tr>
            <tr><td><i>–ö—É—Ä—Å–∏–≤</i></td><td><code>__–ö—É—Ä—Å–∏–≤–Ω—ã–π —Ç–µ–∫—Å—Ç__</code></td></tr>
            <tr><td><u>–ü–æ–¥—á–µ—Ä–∫–Ω—É—Ç—ã–π</u></td><td><code>^^–ü–æ–¥—á–µ—Ä–∫–Ω—É—Ç—ã–π —Ç–µ–∫—Å—Ç^^</code></td></tr>
            <tr><td><s>–ó–∞—á–µ—Ä–∫–Ω—É—Ç—ã–π</s></td><td><code>~~–ó–∞—á–µ—Ä–∫–Ω—É—Ç—ã–π —Ç–µ–∫—Å—Ç~~</code></td></tr>
            <tr><td><span style='background-color: #555; color: #555;'>–°–ø–æ–π–ª–µ—Ä</span></td><td><code>||–°–∫—Ä—ã—Ç—ã–π —Ç–µ–∫—Å—Ç||</code></td></tr>
            <tr><td><code>–ú–æ–Ω–æ—à–∏—Ä–∏–Ω–Ω—ã–π</code></td><td><code>`–º–æ–Ω–æ—à–∏—Ä–∏–Ω–Ω—ã–π —Ç–µ–∫—Å—Ç`</code></td></tr>
        </table>
        <h4>–°—Å—ã–ª–∫–∏ –∏ –∫–∞—Å—Ç–æ–º–Ω—ã–µ —ç–º–æ–¥–∑–∏</h4>
        <ul>
            <li><b>–û–±—ã—á–Ω–∞—è —Å—Å—ã–ª–∫–∞:</b> <code>[–≤–∏–¥–∏–º—ã–π —Ç–µ–∫—Å—Ç](URL-–∞–¥—Ä–µ—Å)</code><br><i>–ü—Ä–∏–º–µ—Ä:</i> <code>[–°–∞–π—Ç Qt](https://qt.io/)</code></li>
            <li><b>–ö–∞—Å—Ç–æ–º–Ω—ã–π —ç–º–æ–¥–∑–∏:</b> <code>[—ç–º–æ–¥–∑–∏](emoji:ID)</code><br><i>–ü—Ä–∏–º–µ—Ä:</i> <code>[üíô](emoji:5278229754099540071)</code></li>
        </ul>
        <h4>–ë–ª–æ—á–Ω—ã–µ —ç–ª–µ–º–µ–Ω—Ç—ã</h4>
        <ul>
            <li><b>–¶–∏—Ç–∞—Ç–∞:</b> –¢–µ–∫—Å—Ç –∑–∞–∫–ª—é—á–∞–µ—Ç—Å—è –≤ <code>&gt;</code> –∏ <code>&lt;</code>.<br><i>–ü—Ä–∏–º–µ—Ä:</i> <code>&gt;–≠—Ç–æ —Ü–∏—Ç–∞—Ç–∞.&lt;</code></li>
            <li><b>–°–≤–æ—Ä–∞—á–∏–≤–∞–µ–º–∞—è —Ü–∏—Ç–∞—Ç–∞:</b> –¢–µ–∫—Å—Ç –∑–∞–∫–ª—é—á–∞–µ—Ç—Å—è –≤ <code>&gt;^</code> –∏ <code>&lt;^</code>.</li>
            <li><b>–ë–ª–æ–∫ –∫–æ–¥–∞:</b> –¢–µ–∫—Å—Ç –∑–∞–∫–ª—é—á–∞–µ—Ç—Å—è –≤ —Ç—Ä–æ–π–Ω—ã–µ –æ–±—Ä–∞—Ç–Ω—ã–µ –∫–∞–≤—ã—á–∫–∏ <code>```</code>. –ú–æ–∂–Ω–æ —É–∫–∞–∑–∞—Ç—å —è–∑—ã–∫.<br><i>–ü—Ä–∏–º–µ—Ä:</i> <code>```cpp<br>#include &lt;iostream&gt;<br>```</code></li>
        </ul>
        <p><b>–ü—Ä–∞–≤–∏–ª–∞ –≤–ª–æ–∂–µ–Ω–Ω–æ—Å—Ç–∏:</b> –°—Ç–∏–ª–∏ –º–æ–∂–Ω–æ –∫–æ–º–±–∏–Ω–∏—Ä–æ–≤–∞—Ç—å (<code>**__–∂–∏—Ä–Ω—ã–π –∫—É—Ä—Å–∏–≤__**</code>). –ë–ª–æ–∫ –∫–æ–¥–∞ –∏–º–µ–µ—Ç –Ω–∞–∏–≤—ã—Å—à–∏–π –ø—Ä–∏–æ—Ä–∏—Ç–µ—Ç –∏ –æ—Ç–º–µ–Ω—è–µ—Ç –ª—é–±–æ–µ —Ñ–æ—Ä–º–∞—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ –≤–Ω—É—Ç—Ä–∏. –í–Ω—É—Ç—Ä–∏ —Ü–∏—Ç–∞—Ç—ã –Ω–µ–ª—å–∑—è –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –º–æ–Ω–æ—à–∏—Ä–∏–Ω–Ω—ã–π —Å—Ç–∏–ª—å.</p>
    )");
    mainLayout->addWidget(tgHelpBrowser);

    mainLayout->addWidget(new QFrame);

    // --- –ë–ª–æ–∫ 3: –§–æ—Ä–º–∞—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ VK ---
    mainLayout->addWidget(new QLabel("<h3>–§–æ—Ä–º–∞—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ VK (–¥–ª—è —Å—Å—ã–ª–æ–∫):</h3>"));
    QWidget *vkWidget = new QWidget();
    QGridLayout *vkLayout = new QGridLayout(vkWidget);
    addRow(vkLayout, 0, "[https://vk.com/dublyajnaya|–¢–û –î—É–±–ª—è–∂–Ω–∞—è]", "–°—Å—ã–ª–∫–∞ –Ω–∞ –≤–Ω–µ—à–Ω–∏–π —Ä–µ—Å—É—Ä—Å");
    vkLayout->setColumnStretch(1, 1);
    mainLayout->addWidget(vkWidget);

    mainLayout->addStretch(1);

    QVBoxLayout *dialogLayout = new QVBoxLayout(helpDialog);
    dialogLayout->addWidget(scrollArea);
    helpDialog->setLayout(dialogLayout);
    helpDialog->show();
}
